subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/PropertyNode.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.properties;
+
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+
+import javax.annotation.Nullable;
+
+public class PropertyNode extends AbstractBeanNode implements BeanNode<PropertyNode> {
+    private final Object bean;
+    private final ImmutableMap<Object, String> seenBeans;
+
+    public PropertyNode(@Nullable String parentPropertyName, Object bean, ImmutableMap<Object, String> seenBeans, TypeMetadata typeMetadata) {
+        super(parentPropertyName, Preconditions.checkNotNull(bean, "Null is not allowed as nested property '" + parentPropertyName + "'").getClass(), typeMetadata);
+        Preconditions.checkState(!seenBeans.keySet().contains(bean), "Cycles between nested beans are not allowed. Cycle detected between: '%s' and '%s'.", seenBeans.get(bean), parentPropertyName);
+        this.bean = bean;
+        this.seenBeans = seenBeans;
+    }
+
+    public Object getBean() {
+        return bean;
+    }
+
+    public ImmutableMap<Object, String> getSeenBeans() {
+        return ImmutableMap.<Object, String>builder().putAll(seenBeans).put(bean, getParentPropertyName() == null ? "<root>" : getParentPropertyName()).build();
+    }
```

Comment:
This is not how we should handle `seenBeans`. There's a number of problems with the current implementation:

- `ImmutableMap.get()` will use `equals()` to decide if we've seen a bean, so beans that are equal would throw an error here. We should use reference equality instead.
- The way we create a new `ImmutableMap` is inefficient. We should instead have a linked list-style data structure where we can add new heads to the list whenever we go a level deeper. We must check `==` on every element anyway, so using something like a `HashSet` or a `HashMap` has no use.
- We shouldn't store `seenBeans` in the `ProprertyNode`, but instead keep track of it during traversal and pass it in where it's needed.

Reply:
`seenBeans` is now tracked in the context. I now use an Iterable to track the `seenBeans` (or `parentNodes` as they are called now) and compare via `==`.